### PR TITLE
enhancement: continue on basis for future improvements

### PR DESCRIPTION
**Pull request by fschade**
followup pr for https://github.com/fschade/icap-client/pull/1

changes:

fix deprecations
get rid of pointer mutating funcs
remove response parsing from the icap connection
align function flow to be more like happy path

### CODE REVIEW

**Review by 2403905 — APPROVED** [review]

File: client.go [hunk 1]
```diff
@@ -55,13 +47,18 @@ func (c *Client) Do(req *Request) (*Response, error) {
 	req.setDefaultRequestHeaders()
 
 	// convert the request to icap message
-	message, err := toICAPMessage(req)
+	message, err := toICAPRequest(req)
```
Comment (butonic) [inline client.go:50]: IMO this should be a member function of the `req` ...

File: client.go [hunk 2]
```diff
@@ -55,13 +47,18 @@ func (c *Client) Do(req *Request) (*Response, error) {
...
 	if err != nil {
 		return nil, err
 	}
 
 	// send the icap message to the server
-	resp, err := c.conn.Send(message)
+	dataRes, err := c.conn.Send(message)
```
Comment (butonic) [inline client.go:56]: I see that you went the connection to only deal with []byte. In and out ... 
I would prefer a stricter public API that takes Request and Response objects ... but we should maybe mimic whatever the std http package does.

File: conn.go [hunk 3]
```diff
@@ -93,27 +94,22 @@ func (c *ICAPConn) Send(in []byte) (*Response, error) {
...
 			// explicitly breaking because the Read blocks for 100 continue message
 			// fixMe: still unclear why this is happening, find out and fix it
-			if string(data) == icap100ContinueMsg {
+			if bytes.Equal(data, []byte(icap100ContinueMsg)) {
 				break
 			}
 
 			// EOF detected, 0 Double crlf indicates the end of the message
-			if strings.HasSuffix(string(data), "0\r\n\r\n") {
+			if bytes.HasSuffix(data, []byte("0\r\n\r\n")) {
```
Comment (butonic) [inline conn.go:102]: I saw that tha first data block we read is 1096 bytes max. Is that part of the spec?
AFAICT ICAP is prefixing http with its own `\r\n` delimited headers.
Should we mimic what http does and parse the lines as key value and put them into a response struct? um maybe that is what the Response struct does 👀

**Review by butonic — APPROVED** [review]

File: client.go (hunk 1 above)
Comment (fschade) [reply client.go:50]: good point for future pr, i've decided to make small changes

File: client.go (hunk 2 above)
Comment (fschade) [reply client.go:56]: yup, same as above, future pr

File: conn.go (hunk 3 above)
Comment (fschade) [reply conn.go:102]: it was that way in the upstream library, i have to check the RFC to answer, at the moment i try to simplify every part, add tests and then do changes like the one you mention

**Pull request merged**